extension.js

```diff
@@ -41,104 +73,337 @@ export default class PipOnTop extends Extension
+    this._scheduleCheck(window);
   }
 
-  _onWindowRemoved(workspace, window)
-  {
-    if (window._notifyPipTitleId) {
-      window.disconnect(window._notifyPipTitleId);
-      window._notifyPipTitleId = null;
-    }
-    if (window._isPipAble)
+  _onWindowRemoved(workspace, window) {
+    if (!window) return;
+
+    // disconnect notify handler if present
+    this._cleanupNotifyId(window);
+
+    // mark not pip-able
+    try {
       window._isPipAble = null;
+    } catch (e) {
+      // ignore
+    }
+  }
+
+  _cleanupNotifyId(window) {
+    try {
+      if (!window) return;
+      if (window._notifyPipTitleId) {
+        try {
+          window.disconnect(window._notifyPipTitleId);
+        } catch (e) {
+          // disconnect may fail if already disconnected - swallow
+        }
+        window._notifyPipTitleId = null;
+      }
+      // clear any stored wrapper
+      if (window._notifyPipHandler) {
+        window._notifyPipHandler = null;
+      }
+      // cancel scheduled idle if any
+      if (window._pipIdleId) {
+        try {
+          GLib.source_remove(window._pipIdleId);
+        } catch (e) {
+          // ignore
+        }
+        window._pipIdleId = null;
+      }
+    } catch (e) {
+      logError(e, 'PiP on Top: _cleanupNotifyId error');
+    }
+  }
+
+  _scheduleCheck(window) {
+    if (!window) return;
+
+    // if a check is already scheduled, keep it (prevents flooding)
+    if (window._pipIdleId) return;
+
+    try {
+      // schedule a short timeout to avoid racing window destruction
+      window._pipIdleId = GLib.timeout_add(GLib.PRIORITY_DEFAULT, CHECK_DELAY_MS, () => {
+        // clear scheduled id immediately
+        try {
+          window._pipIdleId = null;
+        } catch (e) {
+          // ignore
+        }
+
+        if (!window) return GLib.SOURCE_REMOVE;
+        if (window._destroyed) {
+          // clean up listeners for destroyed window
+          this._cleanupNotifyId(window);
+          return GLib.SOURCE_REMOVE;
+        }
+
+        try {
+          this._checkTitle(window);
+        } catch (e) {
+          logError(e, 'PiP on Top: scheduled _checkTitle error');
+        }
+        return GLib.SOURCE_REMOVE;
+      });
+    } catch (e) {
+      // if scheduling failed, run synchronously but guarded
+      try {
+        this._checkTitle(window);
+      } catch (err) {
+        logError(err, 'PiP on Top: fallback _checkTitle error');
+      }
+    }
+  }
+
+  _isWindowTitlePipLike(title) {
+    if (!title) return false;
+
+    try {
+      if (title === 'Picture-in-Picture') return true;
+      if (title === _('Picture-in-Picture')) return true;
+      if (title === 'Picture in picture') return true;
+      if (title === 'Picture-in-picture') return true;
+      if (title.endsWith && title.endsWith(' - PiP')) return true;
+      if (title === 'TelegramDesktop') return true;
+    } catch (e) {
+      // title could be weird; treat as not PiP
+    }
+    return false;
   }
```

Comment:
Not blocking (I don't even have this sort of power), but why a try block?
What about making it case insensitive and checking if its a string first too?
 
```suggestion
  _isWindowTitlePipLike(title) {
    if (typeof title !== 'string' || !title) return false;
    if (title.endsWith(' - PiP')) return true;
    const lowercased = title.toLowerCase();
    
    const pipLikeStrings = ['picture-in-picture', _('Picture-in-Picture'), 'picture in picture'];
    const appSpecific = ['TelegramDesktop'];
    const searchArray = [...pipLikeStrings, ...appSpecific];
   
    return searchArray.some(( it ) => it === lowercased);
  }
```